### ISSUE TITLE
Confusing error message for associated const

### ISSUE THREAD

**Issue by valff** [rust-lang/rust#56589]
```rust
pub struct Val<T: Foo>(T);

pub trait Foo: Sized {}

pub trait Bar {
    type Int;
    const DEFAULT: Self::Int;
}

impl<T: Foo> Bar for Val<T> {
    type Int = usize;
    const DEFAULT: usize = 0;
}
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=905d243c05ba76f76cb62a175a5a993d))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0277]: the trait bound `T: Foo` is not satisfied
  --> src/lib.rs:12:5
   |
12 |     const DEFAULT: usize = 0;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Foo` is not implemented for `T`
   |
   = help: consider adding a `where T: Foo` bound
   = note: required because of the requirements on the impl of `Bar` for `Val<T>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: Could not compile `playground`.

To learn more, run the command again with --verbose.

```

The error message suggests to add a bound that already exists.

**Comment by Palladinium** [comment]
I'm trying to debug the same error and I can't figure out how to make it compile.

Is the example code valid but the compiler is incorrectly rejecting it, or is this purely a misleading error message hiding a different error?
If the latter is the case, how can one fix the example code to compile?

**Comment by Palladinium** [comment]
After some searching it looks to me like this may be related or a duplicate of #43408.

**Comment by estebank** [comment]
This now compiles in the current beta.
